keps/sig-multicluster/4322-cluster-inventory/README.md

```diff
@@ -0,0 +1,993 @@
+particular command or function in one tool, they can expect the same
+command or function to work in the same way in another tool. This can
+further enhance the usability and adoption of different cluster
+manager.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We try to summarize the most common properties that should be able to
+represent a cluster and support the above use case at the minimum scope.
+
+The target consumers of the API are users who manage the clusters and
+other tools to understand the clusters concept from the API for
+multicluster scheduling, workload distribution and cluster management.
+The API aims to provide the information for the consumers to answer the
+below questions:
+
+* Is the cluster under management?
+* How can I select a cluster?
+* Is the cluster healthy?
+* Does the cluster have certain capabilities or properties?
+* Does the cluster have sufficient resources?
+
+### Cluster Name
+
+It is required that cluster name is unique for each cluster, and it
+should also be unique among different providers (cluster manager). There
+are two option to ensure it is uniqueness
+
+#### Option 1
+
+The `metadata.name` of the cluster should never be set upon creation.
+Only `metadata.generateName` can be set.
+
+#### Option 2
+
+The `metadata.name` and `metadata.generateName` must have prefix which
+should be the same as the `spec.clusterManager.name`. Different cluster
+manager must set a different value of `spec.clusterManager.name` when
+the cluster is created.
+
+These two options can be achieved by a validating webhook or a
+`ValidatingAdmissionPolicy` resource.
+
+### Spec
+
+#### Display name
+
+It is a human-readable name of the cluster set by the consumer of the
+cluster.
+
+#### Cluster Manager
+
+An immutable field set by a cluster manager when this cluster resources
+is created by it. Each cluster manager instance should set a different
+values to this field.
+
+In addition, a predefined label with key "x-k8s.io/cluster-manager" needs
```

Comment:
See if you can make a validating admission policy to enforce this. Recall that you can use CRD validation on fields, but not on labels.

Reply:
Sure. CEL expressions could help make sure this field is immutable.

Actually, there are other purposes that we want to have this **OPTIONAL** label.

- With this label, it will be easier for multiple cluster managers to query the clusters and get notified on clusters that they are responsible. Although this is doable with lister indexers/filtering funcs, labels are handy and quite helpful.
- This label could also serve as a default setting of cluster manager.